### PR TITLE
CODETOOLS-7903560: JMH: async profiler reports only the last fork result

### PR DESCRIPTION
**Pull request by yawkat**
Without this change, async-profiler would write the same result in each fork, overwriting the previous result.

This patch introduces a new "FileResult" class that collects the output paths of each benchmark. In `afterTrial`, the output files are moved to a new path that includes the pid (fork number would be an alternative but it's not as easily available).

At the moment, results of different runs are not coalesced. But with these changes, it would be easy to coalesce at least the results of the "collapsed" output mode. I plan to do this in another PR.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [CODETOOLS-7903560](https://bugs.openjdk.org/browse/CODETOOLS-7903560): JMH: async profiler reports only the last fork result (**Enhancement** - P4)

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jmh.git pull/121/head:pull/121` \
`$ git checkout pull/121`

Update a local copy of the PR: \
`$ git checkout pull/121` \
`$ git pull https://git.openjdk.org/jmh.git pull/121/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 121`

View PR using the GUI difftool: \
`$ git pr show -t 121`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jmh/pull/121.diff">https://git.openjdk.org/jmh/pull/121.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jmh/pull/121#issuecomment-1737240142)

### CODE REVIEW

**Comment by yawkat** [comment]
do i need to do anything for the OCA since i'm an oracle employee?

**Comment by shipilev** [comment]
Also, go to https://github.com/yawkat/jmh/actions and enable testing.

**Comment by shipilev** [comment]
I'll take a look later, but I think new files miss license headers. If you run Maven build, the license plugin is supposed to drop some headers there, which would need some copyright year adjustments. When you push, that would also trigger GHA run, which might reveal some problems. 

Merge from master to get JDK 21 as test config.

**Comment by shipilev** [comment]
Not now, bot.

File: jmh-core-it/src/test/java/org/openjdk/jmh/it/profilers/AsyncProfilerForkTest.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (c) 2014, 2015, Oracle and/or its affiliates. All rights reserved.
```
Comment (shipilev) [inline jmh-core-it/src/test/java/org/openjdk/jmh/it/profilers/AsyncProfilerForkTest.java:2]: There and later, the dates are not correct (I think our current template is not current), please adjust these by hand.

**Review by shipilev — COMMENTED** [review]
Sorry for long delay. This looks fine with a few nits. 

I think we still want to emit the pid-less file. It would help the scenario when we re-run the benchmark over and over again, and just hit "Refresh" on the same async-profile result. If the file name would be different each time due to PID, it would break this workflow. The last file in the sequence we can have both copy with the PID filename, and leave the original behind.

**Comment by yawkat** [comment]
ok ive changed it to copy instead of move.

**Comment by shipilev** [comment]
I am seeing some `WinPerfAsm` failures, these do not look related to this PR. I'll try and fix them separately.

**Review by shipilev — COMMENTED** [review]
Generally looks good.

Infrastructure updates landed in current master, please merge from it to get cleaner GHA runs.

I also suggest we pull in async-profiler to Linux runner and run the new tests with it. Add this to `.github/workflows/pre-integration.yml` after `Set up perf (Linux)` step:

```

    - name: Set up async-profiler (Linux)
      run: |
        curl -L https://github.com/async-profiler/async-profiler/releases/download/v3.0/async-profiler-3.0-linux-x64.tar.gz | tar xzf -
        echo "LD_LIBRARY_PATH=$LD_LIBRARY_PATH:`pwd`/async-profiler-3.0-linux-x64/lib/" >> $GITHUB_ENV
      if: (runner.os == 'Linux')
```

I'll approve once GHA is green.

**Comment by shipilev** [comment]
Aha, see, test failures :) I wonder what are those.

**Comment by yawkat** [comment]
oops i think i forgot to adjust AsyncProfilerTest

**Review by shipilev — APPROVED** [review]
Seems like the basic tests have passed. I am pretty sure the rest would pass. Once GHA is fully green, feel free to integrate.

**Comment by yawkat** [comment]
/summary
Without this change, async-profiler would write the same result in each fork, overwriting the previous result.

This patch introduces a new "FileResult" class that collects the output paths of each benchmark. In afterTrial, the output files are copied to a new path that includes the pid (fork number would be an alternative but it's not as easily available).

**Comment by yawkat** [comment]
Thanks!

**Pull request closed without merging**